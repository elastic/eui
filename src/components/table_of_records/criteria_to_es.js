import { isString } from '../../services/predicate';
import { Query as DefaultQuery } from './search';
import { SortDirection } from '../../services/sort';

export const criteriaToESRequestBody = (criteria, options = {}) => {

  const Query = options.Query || DefaultQuery;

  const body = {};

  if (!criteria.page) {
    body.size = options.size || 10;
    body.from = 0;
  } else {
    body.size = criteria.page.size;
    body.from = body.size * criteria.page.index;
  }

  if (criteria.sort) {
    body.sort = [
      { [criteria.sort.field]: SortDirection.isAsc(criteria.sort.direction) ? 'asc' : 'desc' },
      '_score'
    ];
  }

  if (!criteria.search) {
    body.query = { match_all: {} };
  } else {
    const euiQuery = isString(criteria.search.query) ?
      Query.parse(criteria.search.query) :
      criteria.search.query;
    body.query = euiQuery.toESQuery(options);
  }

  return body;
};
