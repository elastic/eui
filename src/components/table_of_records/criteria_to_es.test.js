import { criteriaToESRequestBody } from './criteria_to_es';

describe('criteriaToESRequestBody', () => {

  test('empty criteria', () => {
    const criteria = {};
    const body = criteriaToESRequestBody(criteria);
    expect(body).toMatchSnapshot();
  });

  test('pagination', () => {
    const criteria = {
      page: {
        size: 5,
        index: 3
      }
    };
    const body = criteriaToESRequestBody(criteria);
    expect(body).toMatchSnapshot();
  });

  test('sorting - asc', () => {
    const criteria = {
      sort: {
        field: 'foo',
        direction: 'asc'
      }
    };
    const body = criteriaToESRequestBody(criteria);
    expect(body).toMatchSnapshot();
  });

  test('sorting - desc', () => {
    const criteria = {
      sort: {
        field: 'foo',
        direction: 'desc'
      }
    };
    const body = criteriaToESRequestBody(criteria);
    expect(body).toMatchSnapshot();
  });

  test('search query', () => {
    const Query = {
      parse(query) {
        return {
          toESQuery() {
            return { query };
          }
        };
      }
    };
    const criteria = {
      search: {
        query: 'this is a query'
      }
    };
    const body = criteriaToESRequestBody(criteria, { Query });
    expect(body).toMatchSnapshot();
  });

  test('pagination, sorting and query', () => {
    const Query = {
      parse(query) {
        return {
          toESQuery() {
            return { query };
          }
        };
      }
    };
    const criteria = {
      page: {
        index: 2,
        size: 8
      },
      sort: {
        field: 'foo',
        direction: 'desc'
      },
      search: {
        query: 'this is a query'
      }
    };
    const body = criteriaToESRequestBody(criteria, { Query });
    expect(body).toMatchSnapshot();
  });
});
