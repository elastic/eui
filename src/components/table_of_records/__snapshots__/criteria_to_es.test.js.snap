// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`criteriaToESRequestBody empty criteria 1`] = `
Object {
  "from": 0,
  "query": Object {
    "match_all": Object {},
  },
  "size": 10,
}
`;

exports[`criteriaToESRequestBody pagination 1`] = `
Object {
  "from": 15,
  "query": Object {
    "match_all": Object {},
  },
  "size": 5,
}
`;

exports[`criteriaToESRequestBody pagination, sorting and query 1`] = `
Object {
  "from": 16,
  "query": Object {
    "query": "this is a query",
  },
  "size": 8,
  "sort": Array [
    Object {
      "foo": "desc",
    },
    "_score",
  ],
}
`;

exports[`criteriaToESRequestBody search query 1`] = `
Object {
  "from": 0,
  "query": Object {
    "query": "this is a query",
  },
  "size": 10,
}
`;

exports[`criteriaToESRequestBody sorting - asc 1`] = `
Object {
  "from": 0,
  "query": Object {
    "match_all": Object {},
  },
  "size": 10,
  "sort": Array [
    Object {
      "foo": "asc",
    },
    "_score",
  ],
}
`;

exports[`criteriaToESRequestBody sorting - desc 1`] = `
Object {
  "from": 0,
  "query": Object {
    "match_all": Object {},
  },
  "size": 10,
  "sort": Array [
    Object {
      "foo": "desc",
    },
    "_score",
  ],
}
`;
